zoneforge/authentication.py

```diff
@@ -0,0 +1,212 @@
+            return {"message": str(error)}, 400
+
+class LoginResource(Resource):
+    def post(self):
+        try:
+            args = login_parser.parse_args()
+            username = args.get('username')
+            password = args.get('password').encode(encoding="utf-8")
+
+            user_entity = db.one_or_404(
+                db.select(model.User).filter_by(username=username),
+                description=f"User '{username}' not found"
+            )
+
+            if not user_entity or not bcrypt.checkpw(password, user_entity.password.encode(encoding="utf-8")):
+                raise Unauthorized("Username and password not match")
+
+            return _generate_token(user_entity.id), 200
+
+        except BadRequest as bad_request:
+            if hasattr(bad_request, "data") and "username" in bad_request.data["errors"]:
+                return {"message": "Missing username field"}, 400
+
+            if hasattr(bad_request, "data") and "password" in bad_request.data["errors"]:
+                return {"message": "Missing password field"}, 400
+
+            return {"message": bad_request.description}, 400
+
+        except NotFound as user_not_found:
+            return {"message": user_not_found.description}, 404
+
+        except Unauthorized as wrong_credentials:
+            return {"message": wrong_credentials.description}, 401
+
+        except Exception as error:
+            return {"message": str(error)}, 400
+
+class RefreshTokenResource(Resource):
+    def post(self):
+        try:
+            args = token_parser.parse_args()
+            token = args.get('Authorization').split(" ")[-1]
+
+            user_refresh_token_data = jwt.decode(
+                token,
+                current_app.config['REFRESH_TOKEN_SECRET'],
+                algorithms="HS256"
+            )
+
+            return _generate_token(user_refresh_token_data["id"]), 200
+
+        except BadRequest as bad_request:
+            if hasattr(bad_request, "data") and "Authorization" in bad_request.data["errors"]:
+                return {"message": "Missing JWT token"}, 400
+
+            return {"message": bad_request.description}, 400
+
+        except Exception as error:
+            return {"message": str(error)}, 400
+        
+class SignupResource(Resource):
+    def post(self):
+        try:
+            args = login_parser.parse_args()
+
+            username = args.get('username')
+            password = args.get('password')
+
+            if not username or len(username) < 3:
+                raise BadRequest("Username must be at least 3 characters long")
+
+            if not password or len(password) < 6:
+                raise BadRequest("Password must be at least 6 characters long")
+
+            user_entity = db.paginate(
+                db.select(model.User).filter_by(username=username),
+                per_page=1
+            )
+
+            if user_entity.items:
+                raise Conflict("Username already exist")
```

Comment:
any reason to not just try querying for the user directly?

```suggestion
            user_exists = db.session.execute(
                select(model.User).filter_by(username=username)
            ).scalar_one_or_none()
            
            if user_exists:
                raise Conflict("Username already exists")
```

Reply:
I tried to use the [Flask_SQLAlchemy queries](https://flask-sqlalchemy.readthedocs.io/en/stable/api/#flask_sqlalchemy.SQLAlchemy.get_or_404) form the document and all return error 404 or needed a Id, as in this case was necessary use search by username, I chose paginate command, but this one is really better option.